File: backend/app/schemas/analytics.py
"""Schemas de analytics."""

from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class SnapshotBase(BaseModel):
    """Schema base de snapshot de producao."""

    plot_id: UUID
    snapshot_date: date
    status: str = "ok"
    health_score: Decimal | None = None
    production_stage: str | None = None
    flowers_per_tree: Decimal | None = None
    total_flowers: int | None = None
    flowering_percentage: Decimal | None = None
    fruits_per_tree: Decimal | None = None
    total_fruits: int | None = None
    avg_fruit_size: Decimal | None = None
    fruit_caliber: str | None = None
    estimated_yield_kg: Decimal | None = None
    estimated_yield_tons: Decimal | None = None
    harvest_start_date: date | None = None
    harvest_end_date: date | None = None
    days_to_harvest: int | None = None
    risk_level: str | None = None
    risk_factors: list[str] = []
    extra_data: dict = {}


class SnapshotCreate(SnapshotBase):
    """Schema para criacao de snapshot."""

    last_soil_reading_id: UUID | None = None
    last_vision_data_id: UUID | None = None


class SnapshotResponse(SnapshotBase):
    """Schema de resposta de snapshot."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    last_soil_reading_id: UUID | None = None
    last_vision_data_id: UUID | None = None
    created_at: datetime
    plot_name: str | None = None
    plot_code: str | None = None


class ProductionAnalyticsResponse(BaseModel):
    """Schema de resposta de analytics de producao."""

    total_plots: int
    plots_with_data: int
    total_fruits: int
    estimated_yield_kg: float
    estimated_yield_tons: float
    status_summary: dict[str, int]
    snapshots: list[dict]


class ForecastResponse(BaseModel):
    """Schema de resposta de previsao de producao."""

    total_estimated_kg: float
    total_estimated_tons: float
    harvest_start: date | None = None
    harvest_end: date | None = None
    plots_ready: int
    plots_in_progress: int


class HistoricalDataPoint(BaseModel):
    """Ponto de dados historicos."""

    date: str
    value: float


class HistoricalDataResponse(BaseModel):
    """Schema de resposta de dados historicos."""

    metric: str
    period: str
    data: list[HistoricalDataPoint]

File: frontend/src/types/analytics.ts
/**
 * Analytics types based on backend schema
 */

export type ProductionStage = 
  | 'floracao' 
  | 'frutificacao' 
  | 'crescimento' 
  | 'maturacao' 
  | 'pronto_colheita';

export type RiskLevel = 'baixo' | 'medio' | 'alto' | 'critico';

export type FruitCaliber = 'pequeno' | 'medio' | 'grande' | 'extra_grande';

export interface PlotProductionSnapshot {
  id: string;
  plot_id: string;
  snapshot_date: string;
  status: 'ok' | 'warning' | 'critical' | 'offline';
  health_score: number | null;
  production_stage: ProductionStage | null;
  flowers_per_tree: number | null;
  total_flowers: number | null;
  flowering_percentage: number | null;
  fruits_per_tree: number | null;
  total_fruits: number | null;
  avg_fruit_size: number | null;
  fruit_caliber: FruitCaliber | null;
  estimated_yield_kg: number | null;
  estimated_yield_tons: number | null;
  harvest_start_date: string | null;
  harvest_end_date: string | null;
  days_to_harvest: number | null;
  risk_level: RiskLevel | null;
  risk_factors: string[];
  last_soil_reading_id: string | null;
  last_vision_data_id: string | null;
  extra_data: Record<string, unknown>;
  created_at: string;
  // Joined fields
  plot_name?: string;
  plot_code?: string;
}

export interface SnapshotCreate {
  plot_id: string;
  snapshot_date: string;
  status?: 'ok' | 'warning' | 'critical' | 'offline';
  health_score?: number;
  production_stage?: ProductionStage;
  flowers_per_tree?: number;
  total_flowers?: number;
  flowering_percentage?: number;
  fruits_per_tree?: number;
  total_fruits?: number;
  avg_fruit_size?: number;
  fruit_caliber?: FruitCaliber;
  estimated_yield_kg?: number;
  estimated_yield_tons?: number;
  harvest_start_date?: string;
  harvest_end_date?: string;
  days_to_harvest?: number;
  risk_level?: RiskLevel;
  risk_factors?: string[];
  extra_data?: Record<string, unknown>;
}

export interface AnalyticsFilters {
  farm_id?: string;
  plot_id?: string;
  start_date?: string;
  end_date?: string;
  production_stage?: ProductionStage;
}

export interface FarmSummary {
  farm_id: string;
  farm_name: string;
  total_area: number | null;
  total_plots: number;
  total_trees: number;
  total_sensors: number;
  sensors_online: number;
  sensors_offline: number;
  plots_ok: number;
  plots_warning: number;
  plots_critical: number;
  plots_offline: number;
  active_alerts: number;
  critical_alerts: number;
  warning_alerts: number;
  avg_moisture: number | null;
  avg_temperature: number | null;
  avg_ph: number | null;
  health_score: number;
  estimated_yield_kg: number;
}
